lib/json-parse.js

```diff
@@ -1,35 +1,48 @@
+const RE2 = require('re2');
+
 module.exports = (txt, reviver, context = 20) => {
+  const regexToken = new RE2(/^Unexpected token.*position\s+(\d+)/i);
+  const regexEndof = new RE2(/^Unexpected end of JSON.*/i);
+
   try {
     return JSON.parse(txt, reviver);
   } catch (e) {
-    if (typeof txt !== 'string') {
-      const isEmptyArray = Array.isArray(txt) && txt.length === 0;
-      const errorMessage = `Cannot parse ${ 
-        isEmptyArray ? 'an empty array' : String(txt)}`;
-      throw new TypeError(errorMessage);
-    }
-    const syntaxErr = e.message.match(/^Unexpected token.*position\s+(\d+)/i);
-    const errIdxBrokenJson = e.message.match(/^Unexpected end of JSON.*/i) ? txt.length - 1 : null;
+    handleJsonNotString(txt);
+    const syntaxErr = e.message.match(regexToken);
+    const errIdxBrokenJson = e.message.match(regexEndof) ? txt.length - 1 : null;
     const errIdx = syntaxErr ? +syntaxErr[1] : errIdxBrokenJson;
-    if (errIdx !== null) {
-      const start = errIdx <= context
-        ? 0
-        : errIdx - context;
-      const end = errIdx + context >= txt.length
-        ? txt.length
-        : errIdx + context;
-      e.message += ` while parsing near '${
-        start === 0 ? '' : '...'
-      }${txt.slice(start, end)}${
-        end === txt.length ? '' : '...'
-      }'`;
-    } else {
-      e.message += ` while parsing '${txt.slice(0, context * 2)}'`;
-    }
+    handleErrIdxNotNull(e, txt, errIdx, context);
     e.offset = errIdx;
     const lines = txt.substr(0, errIdx).split('\n');
     e.startLine = lines.length;
     e.startColumn = lines[lines.length - 1].length;
     throw e;
   }
 };
+
+function handleJsonNotString(txt) {
+  if (typeof txt !== 'string') {
```

Comment:
Here, is possible to change the conditional order and make something like:

```js
if (typeof txt === 'string') {
  return;
}

const isEmptyArray = Array.isArray(txt) && txt.length === 0;
const errorMessage = `Cannot parse ${ 
      isEmptyArray ? 'an empty array' : String(txt)}`;
throw new TypeError(errorMessage);
```

But this is only a suggestion.

Reply:
I did not change the logic when not needed on purpose